crates/client/gateway/src/server/handler.rs

```diff
@@ -0,0 +1,307 @@
+                Err(e) => return e.into(),
+            };
+
+        let status = if Some(block.info.header.block_number) <= last_l1_confirmed_block {
+            BlockStatus::AcceptedOnL1
+        } else {
+            BlockStatus::AcceptedOnL2
+        };
+
+        let block_provider = BlockProvider::new(block, status);
+        create_json_response(hyper::StatusCode::OK, &block_provider)
+    } else if let Ok(block) = MadaraPendingBlock::try_from(block) {
+        let block_provider = PendingBlockProvider::new(block);
+        create_json_response(hyper::StatusCode::OK, &block_provider)
+    } else {
+        GatewayError::InternalServerError.into()
+    }
+}
+
+pub async fn handle_get_state_update(req: Request<Body>, backend: Arc<MadaraBackend>) -> Response<Body> {
+    let params = get_params_from_request(&req);
+    let block_id = match block_id_from_params(&params) {
+        Ok(block_id) => block_id,
+        // Return the error response if the request is malformed
+        Err(e) => return e.into(),
+    };
+
+    let resolved_block_id = match backend
+        .resolve_block_id(&block_id)
+        .or_internal_server_error("resolve_block_id")
+        .and_then(|block_id| block_id.ok_or(StarknetError::block_not_found().into()))
+    {
+        Ok(block_id) => block_id,
+        Err(e) => return e.into(),
+    };
+
+    let state_diff = match backend
+        .get_block_state_diff(&resolved_block_id)
+        .or_internal_server_error("get_block_state_diff")
+        .and_then(|block_id| block_id.ok_or(StarknetError::block_not_found().into()))
+    {
+        Ok(state_diff) => state_diff,
+        Err(e) => return e.into(),
+    };
+
+    let with_block = if include_block_params(&params) {
+        let block = match backend
+            .get_block(&block_id)
+            .or_internal_server_error("get_block")
+            .and_then(|block_id| block_id.ok_or(StarknetError::block_not_found().into()))
+        {
+            Ok(block) => block,
+            Err(e) => return e.into(),
+        };
+        Some(block)
+    } else {
+        None
+    };
+
+    match resolved_block_id.is_pending() {
+        true => {
+            let old_root = match backend
+                .get_block_info(&BlockId::Tag(BlockTag::Latest))
+                .or_internal_server_error("get_block_state_diff")
+            {
+                Ok(Some(block)) => match block.as_nonpending() {
+                    Some(block) => block.header.global_state_root,
+                    None => return GatewayError::InternalServerError.into(),
+                },
+                Ok(None) => Felt::ZERO, // The pending block is actually genesis, so old root is zero
+                Err(e) => return e.into(),
+            };
+            let state_update = PendingStateUpdateProvider { old_root, state_diff: state_diff.into() };
+
+            if let Some(block) = with_block {
+                let block = match MadaraPendingBlock::try_from(block.clone()) {
```

Comment:
use or internal server error here too, we're dropping an error that we really want to show to the console